### PR TITLE
KV pre-exec: don't prefetch indirect state inputs from keys

### PR DESCRIPTION
**Pull request by fabiotudone-da**
## Description

`fetchSubmissionInputs` is currently causing an extra read per submission whenever there are DAML contract key inputs, seriously degrading performance. Quoting @gerolf-da:

> If the state value of a contract key state key changed between fetching inputs and the conflict-detection, then the fingerprint of the contract key state key will have changed as well, and conflict detection should detect that.

I.e., this PR partially reverts #8180.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Review by gerolf-da — APPROVED** [review]

**Comment by fabiotudone-da** [comment]
This fails tests on the very similar #8325, transitioning back to draft so we don't merge it until we figure out what's happening. CC @SamirTalwar-DA

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala
Comment (miklos-da) [inline ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala:48]: Please move this comment somewhere else or remove it -- it is confusing to have such context here in a unit test.

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala
Comment (miklos-da) [inline ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala:46]:
```suggestion
    "read from ledger state only once in presence of (KV) state key inputs of DAML contract keys'" in {
```
```suggestion
    "read from ledger state only once even if a DAML contract key is in the read-set" in {
```

Comment (miklos-da) [inline ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala:46]: Can you move this test case after the happy-path case, please?

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala
Comment (miklos-da) [inline ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala:212]:
```suggestion
```

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala
Comment (miklos-da) [inline ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala:216]:
```suggestion
```

**Comment by fabiotudone-da** [comment]
@gerolf-da @miklos-da could you review the newly added test (and explanation)?

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala
Comment (miklos-da) [inline ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala:45]:
```suggestion
```

**Review by miklos-da — CHANGES_REQUESTED** [review]

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala
Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala:216]: scalafmt did that on save, doesn't it for you?

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala
Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala:212]: scalafmt did that on save, doesn't it for you?

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala
Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala:48]: I think it's important background to keep. It is here because it explains the reason of this test; I could move it to the implementation but the logic there has been removed, so it would not be clear (and I'd rather avoid time references such as "once, we were reading twice here..."). Do you have suggestions about where to move it?

Comment (miklos-da) [reply ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala:48]: Then it's better to remove this test case altogether as it does not help understanding how the code works *currently*. This test had (or rather, would have had) a valid existence before we removed the two-phase retrieval of inputs.

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala
Comment (miklos-da) [reply ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala:212]: I don't have reformatting enabled on save. If the lint checks passed without these whitespaces before, then they are not really needed. The scalafmt settings IntelliJ uses must be aligned with what's being enforced by the lint checks (there seems to be a difference).

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala
Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala:48]: That test was probably [this one](https://github.com/digital-asset/daml/commit/0ee2159218eda085a084ed07ca2d7b245fbb6c8e#diff-2a635860c9f1ce61e8da3014db84dc0358a0d3e8d6da18e3a92d64f6b3935906R96), @SamirTalwar-DA might confirm/rectify. If you think there's no chance we'll have a similar regression in the future I can remove it.

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala
Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala:212]: So apparently there is, I'll rework my setup and remove these NLs.

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala
Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala]: Done but `even` -> `when` because I'm not testing all the cases.

Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala]: Done.

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala
Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala]: Done.

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala
Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala]: Done.

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala
Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala]: Done.

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala
Comment (SamirTalwar-DA) [reply ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala]: Let's purge this test altogether. I don't believe we need a test case to verify we're not doing something.

Comment (SamirTalwar-DA) [reply ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala]: (And let's delete the other one too.)

Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidatorSpec.scala]: Done and added performance-related usage advice ScalaDoc to `StateReader.read`.

**Comment by fabiotudone-da** [comment]
@miklos-da @SamirTalwar-DA can you review and the ScalaDoc addition and approve if it's OK with you?

**Review by miklos-da — APPROVED** [review]

**Pull request merged**
